reactor-core/src/main/java/reactor/core/publisher/FluxSwitchOnFirst.java

```diff
@@ -19,610 +19,1029 @@
+				return;
+			}
+
+			if (hasInboundCancelled(previousState)) {
+				Operators.error(actual,
+						new CancellationException(
+								"FluxSwitchOnFirst has already been cancelled"));
+				return;
+			}
+
+			if (!hasFirstValueReceived(previousState)) {
+				final Throwable t = this.throwable;
+				if (t != null) {
+					Operators.error(actual, t);
+				}
+				else {
+					Operators.complete(actual);
+				}
+				return;
+			}
+
+			this.inner = actual;
+
+			actual.onSubscribe(this);
+
+			previousState = markInboundSubscriberSet(this);
+			if (hasInboundCancelled(previousState)) {
+				actual.onError(
+						new CancellationException(
+								"FluxSwitchOnFirst has already been cancelled"));
+			}
+		}
+
+		@Override
+		boolean tryOnNext(CoreSubscriber<? super T> actual, T t) {
+			actual.onNext(t);
+			return true;
+		}
+	}
+
+	static final class SwitchOnFirstConditionalMain<T, R>
+			extends AbstractSwitchOnFirstMain<T, R>
+			implements Fuseable.ConditionalSubscriber<T> {
+
+		SwitchOnFirstConditionalMain(Fuseable.ConditionalSubscriber<? super R> outer,
+				BiFunction<Signal<? extends T>, Flux<T>, Publisher<? extends R>> transformer,
+				boolean cancelSourceOnComplete) {
+			super(outer, transformer, cancelSourceOnComplete);
+		}
+
+		@Override
+		public void subscribe(CoreSubscriber<? super T> actual) {
+			long previousState = markInboundSubscribedOnce(this);
+			if (hasInboundSubscribedOnce(previousState)) {
+				Operators.error(actual,
+						new IllegalStateException(
+								"FluxSwitchOnFirst allows only one Subscriber"));
+				return;
+			}
+
+			if (hasInboundCancelled(previousState)) {
+				Operators.error(actual,
+						new CancellationException(
+								"FluxSwitchOnFirst has already been cancelled"));
+				return;
+			}
+
+			if (!hasFirstValueReceived(previousState)) {
+				final Throwable t = this.throwable;
+				if (t != null) {
+					Operators.error(actual, t);
+				}
+				else {
+					Operators.complete(actual);
+				}
+				return;
+			}
+
+			this.inner = Operators.toConditionalSubscriber(actual);
+
+			actual.onSubscribe(this);
+
+			previousState = markInboundSubscriberSet(this);
+			if (hasInboundCancelled(previousState)) {
+				Operators.error(actual,
+						new CancellationException(
+								"FluxSwitchOnFirst has already been cancelled"));
+			}
+		}
+
+		@Override
+		@SuppressWarnings("unchecked")
+		public boolean tryOnNext(T t) {
+			if (this.done) {
+				Operators.onNextDropped(t, currentContext());
+				return false;
+			}
+
+			final T f = this.firstValue;
+			if (f == null) {
+				this.firstValue = t;
+
+				final long state = markFirstValueReceived(this);
+				if (hasInboundCancelled(state)) {
+					Operators.onDiscard(t, this.outer.currentContext());
+					return true;
+				}
+
+				final Publisher<? extends R> result;
+				final CoreSubscriber<? super R> o = this.outer;
+
+				try {
+					result = Objects.requireNonNull(this.transformer.apply(Signal.next(t,
```

Comment:
same as above, consider splitting out the `Signal` creation

Reply:
fixed